aten/src/ATen/core/jit_type.h

```diff
@@ -845,36 +845,88 @@ struct CAFFE2_API FutureType : public SingleElementType<TypeKind::FutureType, Fu
   FutureType(TypePtr elem) : SingleElementType(elem) {}
 };
 
+using ::torch::jit::Function;
+struct SerializableType;
+using SerializableTypePtr = std::shared_ptr<SerializableType>;
+
+struct CAFFE2_API SerializableType : public Type {
+  SerializableType(TypeKind tk, c10::optional<c10::QualifiedName> qualifiedName)
+    : Type(tk)
+    , name_(qualifiedName) {}
+
+  std::string python_str() const {
+    TORCH_INTERNAL_ASSERT(name_);
+    return name_->qualifiedName();
+  }
+
+  std::string qualname() const {
+    TORCH_INTERNAL_ASSERT(name_);
+    return name_->qualifiedName();
+  }
+
+  std::string qualifier() const {
+    TORCH_INTERNAL_ASSERT(name_);
+    return name_->prefix();
+  }
+
+  std::string basename() const {
+    TORCH_INTERNAL_ASSERT(name_);
+    return name_->name();
+  }
+
+  const c10::optional<c10::QualifiedName> &qualified_name_obj() const {
+    return name_;
+  }
+
+  bool serializable() const {
+    return name_.has_value();
+  }
+
+  virtual c10::optional<std::string> base_class_name() const = 0;
```

Comment:
This API feels pretty overspecialized to the two classes that are supported here. Interface, for instance, would not be able to work in the API because they do not come with Functions,  just FunctionSchema.  What is stopping us from just having a switch statement in the serializer to handle NamedTuple and ClassType differently?